<!DOCTYPE html>
<html>
<head><title>R: Experimental Design and Power Analysis for Perturb-Seq
Experiments</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> Experimental Design and Power Analysis for Perturb-Seq
Experiments
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;perturbplan&rsquo; version 0.0.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="dot-compute_underspecified_power_efficient.html">.compute_underspecified_power_efficient</a></td>
<td>Internal function for efficient separated power computation using C++</td></tr>
<tr><td style="width: 25%;"><a href="dot-compute_underspecified_power_separated.html">.compute_underspecified_power_separated</a></td>
<td>Internal function for separated power computation</td></tr>
<tr><td style="width: 25%;"><a href="adjusted_cutoff.html">adjusted_cutoff</a></td>
<td>Compute the adjusted significance level with either BH or Bonferroni procedure.</td></tr>
<tr><td style="width: 25%;"><a href="BH_cutoff_bisection.html">BH_cutoff_bisection</a></td>
<td>Benjamini–Hochberg cutoff with bisection search (C++ back-end)</td></tr>
<tr><td style="width: 25%;"><a href="calculate_power_grid.html">calculate_power_grid</a></td>
<td>Calculate power grid for app heatmap visualization</td></tr>
<tr><td style="width: 25%;"><a href="compute_distribution_teststat.html">compute_distribution_teststat</a></td>
<td>Compute mean and sd of the score test statistic</td></tr>
<tr><td style="width: 25%;"><a href="compute_power_grid_efficient.html">compute_power_grid_efficient</a></td>
<td>Compute power grid using efficient C++ test statistic computation</td></tr>
<tr><td style="width: 25%;"><a href="compute_power_grid_separated.html">compute_power_grid_separated</a></td>
<td>Compute power grid using separated Monte Carlo approach</td></tr>
<tr><td style="width: 25%;"><a href="compute_power_posthoc.html">compute_power_posthoc</a></td>
<td>Compute power for each perturbation-gene pair</td></tr>
<tr><td style="width: 25%;"><a href="compute_power_posthoc_fixed_fc.html">compute_power_posthoc_fixed_fc</a></td>
<td>Compute power for each perturbation-gene pair with fixed fold change</td></tr>
<tr><td style="width: 25%;"><a href="compute_QC.html">compute_QC</a></td>
<td>Compute QC probability for each enhancer-gene pair</td></tr>
<tr><td style="width: 25%;"><a href="compute_test_stat_clean.html">compute_test_stat_clean</a></td>
<td>Helper function that returns named vector directly (cleaner than original)</td></tr>
<tr><td style="width: 25%;"><a href="compute_zero_prob.html">compute_zero_prob</a></td>
<td>Compute probability mass of NB distribution at zero</td></tr>
<tr><td style="width: 25%;"><a href="FDP_estimate.html">FDP_estimate</a></td>
<td>FDP estimate based on rejection probability.</td></tr>
<tr><td style="width: 25%;"><a href="input_check_library_computation.html">input_check_library_computation</a></td>
<td>Input checking function for library_computation</td></tr>
<tr><td style="width: 25%;"><a href="input_check_posthoc.html">input_check_posthoc</a></td>
<td>Input checking function for compute_power_posthoc</td></tr>
<tr><td style="width: 25%;"><a href="input_check_power_function.html">input_check_power_function</a></td>
<td>Input checking function for power_function</td></tr>
<tr><td style="width: 25%;"><a href="launch_app.html">launch_app</a></td>
<td>Launch the PerturbPlan Shiny App</td></tr>
<tr><td style="width: 25%;"><a href="library_computation.html">library_computation</a></td>
<td>Fit the S-M curve between # mapped reads and # ovserved UMIs.</td></tr>
<tr><td style="width: 25%;"><a href="library_estimation.html">library_estimation</a></td>
<td>Compute the average total UMI per cell and UMI variation parameters.</td></tr>
<tr><td style="width: 25%;"><a href="obtain_expression_information.html">obtain_expression_information</a></td>
<td>Obtain gene-level expression and filtering information based on TPM.</td></tr>
<tr><td style="width: 25%;"><a href="obtain_mapping_efficiency.html">obtain_mapping_efficiency</a></td>
<td>Compute mapping efficiency from QC'd molecule info and Cell Ranger metrics.</td></tr>
<tr><td style="width: 25%;"><a href="obtain_qc_h5_data.html">obtain_qc_h5_data</a></td>
<td>Obtain a data frame with information of all QC'd reads for library estimation.</td></tr>
<tr><td style="width: 25%;"><a href="obtain_qc_response_data.html">obtain_qc_response_data</a></td>
<td>Load and QC gene expression matrix from Cell Ranger .mtx format.</td></tr>
<tr><td style="width: 25%;"><a href="obtain_random_pairs.html">obtain_random_pairs</a></td>
<td>Generate random gRNA–gene discovery pairs for control or simulation.</td></tr>
<tr><td style="width: 25%;"><a href="power_function.html">power_function</a></td>
<td>Compute approximate power of a CRISPR screen</td></tr>
<tr><td style="width: 25%;"><a href="rejection_computation.html">rejection_computation</a></td>
<td>Compute the rejection probability.</td></tr>
<tr><td style="width: 25%;"><a href="summary_h5_data.html">summary_h5_data</a></td>
<td>Obtain the summary statistics of the QC'd data.</td></tr>
<tr><td style="width: 25%;"><a href="var_nb.html">var_nb</a></td>
<td>Variance of NB distribution</td></tr>
</table>
</div></body></html>
