<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="perturbplan">
<title>Prospective Power Analysis Beyond the Web App • perturbplan</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prospective Power Analysis Beyond the Web App">
<meta property="og:description" content="perturbplan">
<meta property="og:image" content="https://katsevich-lab.github.io/perturbplan/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">perturbplan</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/preprocess-reference.html">Prepare Data For Web App</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://katsevich-lab-perturbplan.share.connect.posit.cloud/">
    <span class="fa fa-external-link-alt"></span>
     
    Web App
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Katsevich-Lab/perturbplan" aria-label="GitHub repository">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Prospective Power Analysis Beyond the Web App</h1>
            
      
      
      <div class="d-none name"><code>prospective-power.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates advanced prospective power analysis
workflows that go beyond the web application’s built-in functionality.
We’ll explore:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Comparing perturb-seq and TAP-seq technologies</strong> for
the same biological question</li>
<li>
<strong>Comparing pre-existing experimental designs</strong> instead
of designing from scratch</li>
<li>
<strong>Visualizing expression distributions and saturation
curves</strong> to understand the underlying models</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://katsevich-lab.github.io/perturbplan">perturbplan</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="comparing-perturb-seq-and-tap-seq-technologies">1. Comparing Perturb-seq and TAP-seq Technologies<a class="anchor" aria-label="anchor" href="#comparing-perturb-seq-and-tap-seq-technologies"></a>
</h2>
<p>One common question is whether to use standard perturb-seq (whole
transcriptome) or TAP-seq (targeted amplification and sequencing) for a
specific experiment. We can compare these technologies using the two
pilot datasets: the Gasperini et al. (2019) perturb-seq data and the Ray
et al. (2025) TAP-seq data.</p>
<div class="section level3">
<h3 id="load-pilot-data">Load Pilot Data<a class="anchor" aria-label="anchor" href="#load-pilot-data"></a>
</h3>
<p>First, let’s load the K562 perturb-seq pilot data and the K562
TAP-seq pilot data that comes with the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load K562 pilot data</span></span>
<span><span class="va">pilot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pilot_data_from_package.html">get_pilot_data_from_package</a></span><span class="op">(</span><span class="st">"K562"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract components</span></span>
<span><span class="va">baseline_expression</span> <span class="op">&lt;-</span> <span class="va">pilot_data</span><span class="op">$</span><span class="va">baseline_expression_stats</span></span>
<span><span class="va">library_params</span> <span class="op">&lt;-</span> <span class="va">pilot_data</span><span class="op">$</span><span class="va">library_parameters</span></span>
<span><span class="va">mapping_efficiency</span> <span class="op">&lt;-</span> <span class="va">pilot_data</span><span class="op">$</span><span class="va">mapping_efficiency</span></span>
<span></span>
<span><span class="co"># View data structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">baseline_expression</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    19942 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ response_id        : chr  "ENSG00000243485" "ENSG00000238009" "ENSG00000241860" "ENSG00000286448" ...</span></span>
<span><span class="co">#&gt;  $ relative_expression: num  1.50e-07 1.55e-06 7.33e-07 9.31e-07 7.43e-06 ...</span></span>
<span><span class="co">#&gt;  $ expression_size    : num  0.3556 2.4156 7.7329 0.0341 3.4513 ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Number of genes:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">baseline_expression</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of genes: 19942</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">library_params</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ UMI_per_cell: num 59163</span></span>
<span><span class="co">#&gt;  $ variation   : num 0.397</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">mapping_efficiency</span><span class="op">)</span></span>
<span><span class="co">#&gt;  num 0.711</span></span>
<span></span>
<span><span class="co"># load the K562 TAP-seq pilot data</span></span>
<span><span class="va">tapseq_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"K562_Ray"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract TAP-seq components</span></span>
<span><span class="va">tapseq_baseline_expression</span> <span class="op">&lt;-</span> <span class="va">K562_Ray</span><span class="op">$</span><span class="va">baseline_expression_stats</span></span>
<span><span class="va">tapseq_library_params</span> <span class="op">&lt;-</span> <span class="va">K562_Ray</span><span class="op">$</span><span class="va">library_parameters</span></span>
<span><span class="va">tapseq_mapping_efficiency</span> <span class="op">&lt;-</span> <span class="va">K562_Ray</span><span class="op">$</span><span class="va">mapping_efficiency</span></span>
<span></span>
<span><span class="co"># View TAP-seq data structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tapseq_baseline_expression</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    303 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ response_id        : chr  "ENSG00000235169" "ENSG00000130764" "ENSG00000116198" "ENSG00000159023" ...</span></span>
<span><span class="co">#&gt;  $ relative_expression: num  0.003732 0.000337 0.000121 0.001622 0.00034 ...</span></span>
<span><span class="co">#&gt;  $ expression_size    : num  2.55 3.92 18.58 5.21 1.73 ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Number of genes:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tapseq_baseline_expression</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of genes: 303</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tapseq_library_params</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ UMI_per_cell: num 2377</span></span>
<span><span class="co">#&gt;  $ variation   : num 0.809</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">tapseq_mapping_efficiency</span><span class="op">)</span></span>
<span><span class="co">#&gt;  num 0.349</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="key-differences-between-technologies">Key Differences Between Technologies<a class="anchor" aria-label="anchor" href="#key-differences-between-technologies"></a>
</h3>
<p>The two technologies differ in several important ways:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Gene coverage</strong>: Perturb-seq measures ~20,000 genes
genome-wide, while TAP-seq focuses on a targeted panel of ~300
genes</li>
<li>
<strong>Mapping efficiency</strong>: TAP-seq has higher mapping
efficiency to the target gene panel (~70%) compared to perturb-seq
(~30%)</li>
<li>
<strong>Library parameters</strong>: TAP-seq shows lower
UMI_per_cell and higher variation in PCR amplification compared to
perturb-seq</li>
</ol>
</div>
<div class="section level3">
<h3 id="power-comparison">Power Comparison<a class="anchor" aria-label="anchor" href="#power-comparison"></a>
</h3>
<p>Suppose we use TAP-seq and perturb-seq to analyze the same set of
pairs targeting at the TAP-seq genes. Let’s first compare how the
overall power scales with sequencing depth for both technologies, fixing
the number of cells at 2,000 per gRNA target:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define read depth range</span></span>
<span><span class="va">read_depths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span>, <span class="fl">1600000</span><span class="op">)</span></span>
<span><span class="va">cells_per_target</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span></span>
<span><span class="co"># Experimental parameters</span></span>
<span><span class="va">MOI</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">num_targets</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">minimum_fold_change</span> <span class="op">&lt;-</span> <span class="fl">0.8</span></span>
<span><span class="va">prop_non_null</span> <span class="op">&lt;-</span> <span class="fl">0.005</span></span>
<span><span class="va">multiple_testing_alpha</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">TPM_threshold</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span></span>
<span><span class="co"># Gene list we consider </span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="va">tapseq_baseline_expression</span><span class="op">$</span><span class="va">response_id</span></span>
<span></span>
<span><span class="va">comparison_results_reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">read_depths</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">reads</span> <span class="op">&lt;-</span> <span class="va">read_depths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Perturb-seq power</span></span>
<span>  <span class="co"># only keep the genes in the TAP-seq panel</span></span>
<span>  <span class="va">baseline_expression_on_gene_list</span> <span class="op">&lt;-</span> <span class="va">baseline_expression</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">response_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gene_list</span><span class="op">)</span></span>
<span>  <span class="va">power_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan.html">compute_power_plan</a></span><span class="op">(</span></span>
<span>    baseline_expression_stats <span class="op">=</span> <span class="va">baseline_expression_on_gene_list</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">library_params</span>,</span>
<span>    cells_per_target <span class="op">=</span> <span class="va">cells_per_target</span>,</span>
<span>    sequenced_reads_per_cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>    MOI <span class="op">=</span> <span class="va">MOI</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="va">num_targets</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    prop_non_null <span class="op">=</span> <span class="va">prop_non_null</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="va">multiple_testing_alpha</span>,</span>
<span>    TPM_threshold <span class="op">=</span> <span class="va">TPM_threshold</span>,</span>
<span>    minimum_fold_change <span class="op">=</span> <span class="va">minimum_fold_change</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="co"># TAP-seq power</span></span>
<span>  <span class="va">power_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan.html">compute_power_plan</a></span><span class="op">(</span></span>
<span>    baseline_expression_stats <span class="op">=</span> <span class="va">tapseq_baseline_expression</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">tapseq_library_params</span>,</span>
<span>    cells_per_target <span class="op">=</span> <span class="va">cells_per_target</span>,</span>
<span>    sequenced_reads_per_cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>    MOI <span class="op">=</span> <span class="va">MOI</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="va">num_targets</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    prop_non_null <span class="op">=</span> <span class="va">prop_non_null</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="va">multiple_testing_alpha</span>,</span>
<span>    TPM_threshold <span class="op">=</span> <span class="va">TPM_threshold</span>,</span>
<span>    minimum_fold_change <span class="op">=</span> <span class="va">minimum_fold_change</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">comparison_results_reads</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>    Technology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Perturb-seq"</span>, <span class="st">"TAP-seq"</span><span class="op">)</span>,</span>
<span>    Overall_Power <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">power_ps</span><span class="op">$</span><span class="va">overall_power</span>, <span class="va">power_ts</span><span class="op">$</span><span class="va">overall_power</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">comparison_reads_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">comparison_results_reads</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot overall power vs read depth</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">comparison_reads_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">Overall_Power</span>,</span>
<span>                          color <span class="op">=</span> <span class="va">Technology</span>, linetype <span class="op">=</span> <span class="va">Technology</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Technology Comparison: Power vs Sequencing Depth"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s cells per target, FC = %.2f, FDR = %.0f%%"</span>,</span>
<span>                      <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">(</span><span class="va">cells_per_target</span><span class="op">)</span>, <span class="va">minimum_fold_change</span>,</span>
<span>                      <span class="va">multiple_testing_alpha</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Overall Power"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Technology"</span>,</span>
<span>    linetype <span class="op">=</span> <span class="st">"Technology"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>    panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-compare-read-depths-1.png" width="100%"></p>
<p>Then, let’s compare how power scales with the number of cells per
target, fixing sequencing depth at 40,000 reads per cell:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define cell number range</span></span>
<span><span class="va">cell_numbers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">500</span>, <span class="fl">1000</span>, <span class="fl">2000</span>, <span class="fl">4000</span><span class="op">)</span></span>
<span><span class="va">reads_per_cell</span> <span class="op">&lt;-</span> <span class="fl">40000</span></span>
<span></span>
<span><span class="va">comparison_results_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">cell_numbers</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cell_numbers</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Perturb-seq power</span></span>
<span>  <span class="va">power_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan.html">compute_power_plan</a></span><span class="op">(</span></span>
<span>    baseline_expression_stats <span class="op">=</span> <span class="va">baseline_expression_on_gene_list</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">library_params</span>,</span>
<span>    cells_per_target <span class="op">=</span> <span class="va">cells</span>,</span>
<span>    sequenced_reads_per_cell <span class="op">=</span> <span class="va">reads_per_cell</span>,</span>
<span>    MOI <span class="op">=</span> <span class="va">MOI</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="va">num_targets</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    prop_non_null <span class="op">=</span> <span class="va">prop_non_null</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="va">multiple_testing_alpha</span>,</span>
<span>    TPM_threshold <span class="op">=</span> <span class="va">TPM_threshold</span>,</span>
<span>    minimum_fold_change <span class="op">=</span> <span class="va">minimum_fold_change</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="co"># TAP-seq power</span></span>
<span>  <span class="va">power_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan.html">compute_power_plan</a></span><span class="op">(</span></span>
<span>    baseline_expression_stats <span class="op">=</span> <span class="va">tapseq_baseline_expression</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">tapseq_library_params</span>,</span>
<span>    cells_per_target <span class="op">=</span> <span class="va">cells</span>,</span>
<span>    sequenced_reads_per_cell <span class="op">=</span> <span class="va">reads_per_cell</span>,</span>
<span>    MOI <span class="op">=</span> <span class="va">MOI</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="va">num_targets</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    prop_non_null <span class="op">=</span> <span class="va">prop_non_null</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="va">multiple_testing_alpha</span>,</span>
<span>    TPM_threshold <span class="op">=</span> <span class="va">TPM_threshold</span>,</span>
<span>    minimum_fold_change <span class="op">=</span> <span class="va">minimum_fold_change</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">comparison_results_cells</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Cells_per_Target <span class="op">=</span> <span class="va">cells</span>,</span>
<span>    Technology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Perturb-seq"</span>, <span class="st">"TAP-seq"</span><span class="op">)</span>,</span>
<span>    Overall_Power <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">power_ps</span><span class="op">$</span><span class="va">overall_power</span>, <span class="va">power_ts</span><span class="op">$</span><span class="va">overall_power</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">comparison_cells_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">comparison_results_cells</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot overall power vs cell number</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">comparison_cells_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Cells_per_Target</span>, y <span class="op">=</span> <span class="va">Overall_Power</span>,</span>
<span>                          color <span class="op">=</span> <span class="va">Technology</span>, linetype <span class="op">=</span> <span class="va">Technology</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Technology Comparison: Power vs Cell Numbers"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s reads per cell, FC = %.2f, FDR = %.0f%%"</span>,</span>
<span>                      <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">(</span><span class="va">reads_per_cell</span><span class="op">)</span>, <span class="va">minimum_fold_change</span>,</span>
<span>                      <span class="va">multiple_testing_alpha</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Cells per Target"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Overall Power"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Technology"</span>,</span>
<span>    linetype <span class="op">=</span> <span class="st">"Technology"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>    panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-compare-cell-numbers-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h3>
<p><strong>Power differences across sequencing depths.</strong> TAP-seq
consistently achieves higher overall power than perturb-seq at the same
sequencing depth when cell numbers are held constant. This advantage
stems from TAP-seq’s higher mapping efficiency (~70% vs ~30%), which
effectively provides more UMIs mapped to genes of interest per read.
However, both technologies show diminishing returns at higher read
depths due to UMI saturation effects. The power gap between technologies
narrows at very high read depths (&gt;80,000 reads/cell), suggesting
that extreme sequencing depths may not justify TAP-seq’s additional
panel design costs if reads are abundant.</p>
<p><strong>Power differences across cell numbers.</strong> When varying
cell numbers at fixed sequencing depth, TAP-seq maintains its power
advantage, but the relative benefit is smaller than when varying read
depth. Both technologies show roughly square-root scaling of power with
cell numbers (power ∝ √n), as expected from statistical theory. This
suggests that for both technologies, increasing cell numbers provides
consistent but diminishing returns. The similar scaling patterns
indicate that the choice between technologies should primarily be driven
by the sequencing depth regime you plan to operate in, rather than the
number of cells you can collect.</p>
<p><strong>Practical implications.</strong> TAP-seq is most advantageous
when you have well-defined target genes and operate at moderate
sequencing depths (20,000-60,000 reads/cell), where its higher mapping
efficiency translates to substantial power gains. Perturb-seq remains
preferable for exploratory studies where target genes are uncertain, as
it provides genome-wide coverage without requiring panel design. For
validation experiments with known targets at moderate depths, TAP-seq’s
power advantage often justifies the panel design investment. However, if
you can afford very deep sequencing (&gt;80,000 reads/cell), the
technologies converge in performance, and perturb-seq’s flexibility may
outweigh TAP-seq’s marginal power advantage.</p>
</div>
</div>
<div class="section level2">
<h2 id="comparing-pre-existing-experimental-designs">2. Comparing Pre-existing Experimental Designs<a class="anchor" aria-label="anchor" href="#comparing-pre-existing-experimental-designs"></a>
</h2>
<p>Sometimes you want to compare power between experiments that have
already been performed or are constrained by available resources. Let’s
compare four experimental designs:</p>
<div class="section level3">
<h3 id="define-competing-designs">Define Competing Designs<a class="anchor" aria-label="anchor" href="#define-competing-designs"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Three competing designs with similar total sequencing budget</span></span>
<span><span class="va">designs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Deep"</span>, <span class="st">"Balanced"</span>, <span class="st">"Wide"</span>, <span class="st">"Complete"</span><span class="op">)</span>,</span>
<span>  Cells_per_gRNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">5000</span>, <span class="fl">10000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  Reads_per_Cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">80000</span>, <span class="fl">40000</span>, <span class="fl">20000</span>, <span class="fl">80000</span><span class="op">)</span>,</span>
<span>  Description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Fewer cells, deeper sequencing"</span>,</span>
<span>    <span class="st">"Balanced cells and depth"</span>,</span>
<span>    <span class="st">"More cells, shallow sequencing"</span>,</span>
<span>    <span class="st">"More cells, deeper sequencing"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">designs</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Design Cells_per_gRNA Reads_per_Cell                    Description</span></span>
<span><span class="co">#&gt; 1     Deep           2000          80000 Fewer cells, deeper sequencing</span></span>
<span><span class="co">#&gt; 2 Balanced           5000          40000       Balanced cells and depth</span></span>
<span><span class="co">#&gt; 3     Wide          10000          20000 More cells, shallow sequencing</span></span>
<span><span class="co">#&gt; 4 Complete          10000          80000  More cells, deeper sequencing</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-power-for-each-design">Calculate Power for Each Design<a class="anchor" aria-label="anchor" href="#calculate-power-for-each-design"></a>
</h3>
<p>Suppose we’re designing a perturb-seq experiment on K562 cell line.
Suppose we want to perturb 100 targeted genes with the MOI of 10, we
expect a minimum fold change of 0.8 in 10% of the target-gene pairs we
test, and we want to control FDR at 10%. Let’s compare the overall power
and overall cost for each design:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_power_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">designs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span>  <span class="va">power_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan.html">compute_power_plan</a></span><span class="op">(</span></span>
<span>    baseline_expression_stats <span class="op">=</span> <span class="va">baseline_expression</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">library_params</span>,</span>
<span>    cells_per_target <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Cells_per_gRNA</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    sequenced_reads_per_cell <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Reads_per_Cell</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    MOI <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    prop_non_null <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    TPM_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    minimum_fold_change <span class="op">=</span> <span class="fl">0.8</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">design_power_results</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Design <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Design</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    Cells_per_gRNA <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Cells_per_gRNA</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    Num_captured_Cells <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">num_captured_cells</span>,</span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Reads_per_Cell</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    Overall_Power <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">overall_power</span>,</span>
<span>    Overall_Cost <span class="op">=</span> <span class="fu">perturbplan</span><span class="fu">:::</span><span class="fu"><a href="../reference/cost_computation.html">cost_computation</a></span><span class="op">(</span></span>
<span>      num_captured_cells <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">num_captured_cells</span>,</span>
<span>      sequenced_reads_per_cell <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">sequenced_reads_per_cell</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">design_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">design_power_results</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">design_comparison</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Design Cells_per_gRNA Num_captured_Cells Reads_per_Cell Overall_Power</span></span>
<span><span class="co">#&gt; 1     Deep           2000              20500          80000     0.5522817</span></span>
<span><span class="co">#&gt; 2 Balanced           5000              51250          40000     0.6192765</span></span>
<span><span class="co">#&gt; 3     Wide          10000             102500          20000     0.6430014</span></span>
<span><span class="co">#&gt; 4 Complete          10000             102500          80000     0.7466946</span></span>
<span><span class="co">#&gt;   Overall_Cost</span></span>
<span><span class="co">#&gt; 1      2376.36</span></span>
<span><span class="co">#&gt; 2      5174.20</span></span>
<span><span class="co">#&gt; 3      9581.70</span></span>
<span><span class="co">#&gt; 4     11881.80</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-power-comparison">Visualize Power Comparison<a class="anchor" aria-label="anchor" href="#visualize-power-comparison"></a>
</h3>
<p>Let’s visualize the four designs in the two dimensional space of
overall power and cost:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">design_comparison</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Overall_Cost</span>, y <span class="op">=</span> <span class="va">Overall_Power</span>, color <span class="op">=</span> <span class="va">Design</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Power Comparison of Pre-existing Designs"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Total Cost (USD)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Overall Power"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Design"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>    panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-power-comparison-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="interpretation-1">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-1"></a>
</h3>
<p><strong>Non-linear cost-power relationship.</strong> The comparison
reveals important trade-offs between overall cost and overall power
across the four experimental designs. Higher cost generally leads to
higher power, but the relationship is highly non-linear, with critical
implications for experimental planning.</p>
<p>Although the cost differences between consecutive designs are similar
in magnitude, the power gains vary dramatically. The largest power
improvements come from moving away from extreme designs: transitioning
from “Deep” (2,000 cells, 80K reads) to “Balanced” (5,000 cells, 40K
reads) or from “Wide” (10,000 cells, 20K reads) to “Complete” (10,000
cells, 80K reads) yields substantial power gains.</p>
<p><strong>Statistical mechanisms underlying the non-linear
relationship.</strong> The different power gain from the change of
experimental parameters emerges from the interaction between two
distinct statistical effects. First, the saturation curve (explored in
Section 3) shows that UMI counts plateau at high read depths, limiting
the benefit of deeper sequencing beyond ~40,000 reads per cell for this
dataset. Second, statistical power from increased cell numbers follows a
square-root relationship (power scales roughly with √n), meaning
doubling the number of cells provides diminishing marginal returns.</p>
<p><strong>Resource allocation strategies for budget-constrained
experiments.</strong> If budget is limited and you can only afford
incremental increases, prioritize moving away from extreme designs. For
example, if currently at the “Deep” design, adding more cells (moving
toward “Balanced”) typically provides better return on investment than
adding more sequencing depth.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-expression-distributions-and-saturation-curves">3. Visualizing Expression Distributions and Saturation Curves<a class="anchor" aria-label="anchor" href="#visualizing-expression-distributions-and-saturation-curves"></a>
</h2>
<p>In this section, we’ll visualize the K562 perturb-seq pilot data from
Gasperini et. al.(2019) in terms of baseline_expression_stats and
library_params. These visualizations help understand the cases where the
data from Gasperini et. al.(2019) is a good reference for your
experiment and the statistical mechanisms underlying the power analyses
in Sections 1 and 2.</p>
<div class="section level3">
<h3 id="gene-expression-distribution">Gene Expression Distribution<a class="anchor" aria-label="anchor" href="#gene-expression-distribution"></a>
</h3>
<p>From the baseline_expression_stats in the pilot data, let’s visualize
the distribution of relative expression (TPM) and the dispersion size
across genes in the K562 pilot data:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare expression data</span></span>
<span><span class="va">expression_df</span> <span class="op">&lt;-</span> <span class="va">baseline_expression</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    TPM <span class="op">=</span> <span class="va">relative_expression</span> <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span>    Log10_TPM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">TPM</span> <span class="op">+</span> <span class="fl">0.1</span><span class="op">)</span>  <span class="co"># Add pseudocount for log</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Histogram of TPM values</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">expression_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TPM</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span>, fill <span class="op">=</span> <span class="st">"steelblue"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span></span>
<span>    labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span>,</span>
<span>    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">100</span>, <span class="fl">1000</span>, <span class="fl">10000</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Distribution of Gene Expression Levels"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%d genes in baseline data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">expression_df</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Expression Level (TPM)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Number of Genes"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.minor.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"gray90"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-expression-distribution-1.png" width="100%"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Expression vs dispersion relationship</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">expression_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TPM</span>, y <span class="op">=</span> <span class="va">expression_size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, color <span class="op">=</span> <span class="st">"red"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Gene Expression vs Dispersion Parameter"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="st">"Lower dispersion = higher biological variability"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Expression Level (TPM)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Dispersion Size Parameter"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-expression-distribution-2.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="read-saturation-curve-visualization">Read saturation Curve Visualization<a class="anchor" aria-label="anchor" href="#read-saturation-curve-visualization"></a>
</h3>
<p>From the library_params in the pilot data, let’s first visualize the
relationship between reads per cell and UMIs per cell, and see how the
experimental designs in Section 2 map onto the curve:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract library parameters</span></span>
<span><span class="va">UMI_max</span> <span class="op">&lt;-</span> <span class="va">library_params</span><span class="op">$</span><span class="va">UMI_per_cell</span></span>
<span><span class="va">variation</span> <span class="op">&lt;-</span> <span class="va">library_params</span><span class="op">$</span><span class="va">variation</span></span>
<span></span>
<span><span class="co"># Generate saturation curve</span></span>
<span><span class="va">reads_range</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">100000</span>, length.out <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">umis_predicted</span> <span class="op">&lt;-</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">reads_range</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                               <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="va">reads_range</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">saturation_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Reads_per_Cell <span class="op">=</span> <span class="va">reads_range</span>,</span>
<span>  UMIs_per_Cell <span class="op">=</span> <span class="va">umis_predicted</span>,</span>
<span>  Saturation_Pct <span class="op">=</span> <span class="op">(</span><span class="va">umis_predicted</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot saturation curve</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">saturation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">UMIs_per_Cell</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">UMI_max</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">80000</span>, y <span class="op">=</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="fl">1.05</span>,</span>
<span>           label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Max UMI = %d"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">UMI_max</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"red"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Add example design points</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span>,</span>
<span>    Design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Wide"</span>, <span class="st">"Balanced"</span>, <span class="st">"Deep"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">Reads_per_Cell</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                                              <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="va">Reads_per_Cell</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span>,</span>
<span>    Design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Wide"</span>, <span class="st">"Balanced"</span>, <span class="st">"Deep"</span><span class="op">)</span>,</span>
<span>    UMI <span class="op">=</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                       <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">UMI</span>, label <span class="op">=</span> <span class="va">Design</span><span class="op">)</span>,</span>
<span>  vjust <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, color <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Read-to-UMI Saturation Curve"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"UMI_per_cell = %d, variation = %.3f"</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">UMI_max</span><span class="op">)</span>, <span class="va">variation</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"UMIs per Cell"</span>,</span>
<span>    caption <span class="op">=</span> <span class="st">"Green points show example experimental designs"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-saturation-curve-1.png" width="100%"></p>
<p>Also, we can visualize how the percentage of UMI detected
(saturation) increases with read depth:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Saturation percentage vs reads</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">saturation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">Saturation_Pct</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">80</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"orange"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">90</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">90000</span>, y <span class="op">=</span> <span class="fl">82</span>, label <span class="op">=</span> <span class="st">"80% saturation"</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"orange"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">90000</span>, y <span class="op">=</span> <span class="fl">92</span>, label <span class="op">=</span> <span class="st">"90% saturation"</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"red"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Sequencing Saturation vs Read Depth"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Saturation (%)"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="articles/figures/prospective-power-saturation-percentage-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="see-also">See Also<a class="anchor" aria-label="anchor" href="#see-also"></a>
</h2>
<ul>
<li>
<code><a href="../reference/compute_power_plan.html">?compute_power_plan</a></code>: Prospective power analysis for
experimental planning</li>
<li>
<code><a href="../reference/compute_power_plan_per_grid.html">?compute_power_plan_per_grid</a></code>: Grid-based power
analysis</li>
<li>
<code><a href="../reference/library_estimation.html">?library_estimation</a></code>: Library parameter estimation from
QC data</li>
<li>Vignette “Preprocess Reference Expression Data”: Creating custom
pilot data</li>
<li>Vignette “Getting Started with PerturbPlan”: Basic power analysis
workflow</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ziang Niu, Yihui He, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
