<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Prospective Power Analysis Beyond the Web App • perturbplan</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prospective Power Analysis Beyond the Web App">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">perturbplan</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/preprocess-reference.html">Prepare Data For Web App</a></li>
    <li><a class="dropdown-item" href="../articles/prospective-power.html">Prospective Power Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/posthoc.html">Retrospective Power Analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://katsevich-lab-perturbplan.share.connect.posit.cloud/"><span class="fa fa-external-link-alt"></span> Web App</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Katsevich-Lab/perturbplan" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Prospective Power Analysis Beyond the Web App</h1>
            
      

      <div class="d-none name"><code>prospective-power.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates advanced prospective power analysis
workflows that go beyond the web application’s built-in functionality.
We’ll explore:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Comparing perturb-seq and TAP-seq technologies</strong> for
the same biological question</li>
<li>
<strong>Comparing pre-existing experimental designs</strong> instead
of designing from scratch</li>
<li>
<strong>Visualizing expression distributions and saturation
curves</strong> to understand the underlying models</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://katsevich-lab.github.io/perturbplan">perturbplan</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="load-pilot-data">Load Pilot Data<a class="anchor" aria-label="anchor" href="#load-pilot-data"></a>
</h2>
<p>First, let’s load the default K562 pilot data that comes with the
package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load K562 pilot data</span></span>
<span><span class="va">pilot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pilot_data_from_package.html">get_pilot_data_from_package</a></span><span class="op">(</span><span class="st">"K562"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract components</span></span>
<span><span class="va">baseline_expression</span> <span class="op">&lt;-</span> <span class="va">pilot_data</span><span class="op">$</span><span class="va">baseline_expression_stats</span></span>
<span><span class="va">library_params</span> <span class="op">&lt;-</span> <span class="va">pilot_data</span><span class="op">$</span><span class="va">library_parameters</span></span>
<span></span>
<span><span class="co"># View data structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">baseline_expression</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    19942 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ response_id        : chr  "ENSG00000243485" "ENSG00000238009" "ENSG00000241860" "ENSG00000286448" ...</span></span>
<span><span class="co">#&gt;  $ relative_expression: num  1.50e-07 1.55e-06 7.33e-07 9.31e-07 7.43e-06 ...</span></span>
<span><span class="co">#&gt;  $ expression_size    : num  0.3556 2.4156 7.7329 0.0341 3.4513 ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">library_params</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ UMI_per_cell: num 59163</span></span>
<span><span class="co">#&gt;  $ variation   : num 0.397</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="comparing-perturb-seq-and-tap-seq-technologies">1. Comparing Perturb-seq and TAP-seq Technologies<a class="anchor" aria-label="anchor" href="#comparing-perturb-seq-and-tap-seq-technologies"></a>
</h2>
<p>One common question is whether to use standard perturb-seq (whole
transcriptome) or TAP-seq (targeted amplification and sequencing) for a
specific experiment. TAP-seq allows deeper sequencing of a smaller set
of genes, potentially improving power for targeted studies.</p>
<div class="section level3">
<h3 id="simulating-tap-seq-data">Simulating TAP-seq Data<a class="anchor" aria-label="anchor" href="#simulating-tap-seq-data"></a>
</h3>
<p>TAP-seq has two key differences from standard perturb-seq:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Higher mapping efficiency</strong> to the target gene panel
(typically 60-80% vs 20-40%)</li>
<li>
<strong>Focused on a smaller gene set</strong> (e.g., 500-2000 genes
vs whole transcriptome)</li>
</ol>
<p>Let’s create TAP-seq parameters by modifying the perturb-seq
baseline:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select a subset of genes for TAP-seq panel (e.g., top 1000 by expression)</span></span>
<span><span class="va">tapseq_genes</span> <span class="op">&lt;-</span> <span class="va">baseline_expression</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">relative_expression</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">response_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create TAP-seq pilot data (subset of genes with higher mapping efficiency)</span></span>
<span><span class="va">tapseq_pilot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  baseline_expression_stats <span class="op">=</span> <span class="va">baseline_expression</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">response_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">tapseq_genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>relative_expression <span class="op">=</span> <span class="va">relative_expression</span> <span class="op">*</span> <span class="op">(</span><span class="fl">0.70</span> <span class="op">/</span> <span class="fl">0.30</span><span class="op">)</span><span class="op">)</span>,  <span class="co"># Efficiency boost</span></span>
<span>  library_parameters <span class="op">=</span> <span class="va">library_params</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Perturb-seq genes:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">baseline_expression</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Perturb-seq genes: 19942</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"TAP-seq genes:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tapseq_pilot_data</span><span class="op">$</span><span class="va">baseline_expression_stats</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; TAP-seq genes: 1000</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Efficiency boost factor: 0.70 / 0.30 ="</span>, <span class="fl">0.70</span> <span class="op">/</span> <span class="fl">0.30</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Efficiency boost factor: 0.70 / 0.30 = 2.333333</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="power-comparison-single-design-point">Power Comparison: Single Design Point<a class="anchor" aria-label="anchor" href="#power-comparison-single-design-point"></a>
</h3>
<p>Let’s compare power for a specific experimental design using
prospective power analysis:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define experimental parameters</span></span>
<span><span class="va">cells_per_target</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">reads_per_cell</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">minimum_fold_change</span> <span class="op">&lt;-</span> <span class="fl">0.8</span>  <span class="co"># Expected effect size</span></span>
<span><span class="va">gRNA_variability</span> <span class="op">&lt;-</span> <span class="fl">0.13</span>    <span class="co"># Variability between gRNAs</span></span>
<span></span>
<span><span class="co"># Extract fold change and expression info for perturb-seq</span></span>
<span><span class="va">fc_expr_perturbseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_fc_expression_info.html">extract_fc_expression_info</a></span><span class="op">(</span></span>
<span>  minimum_fold_change <span class="op">=</span> <span class="va">minimum_fold_change</span>,</span>
<span>  gRNA_variability <span class="op">=</span> <span class="va">gRNA_variability</span>,</span>
<span>  biological_system <span class="op">=</span> <span class="st">"K562"</span>,</span>
<span>  gene_list <span class="op">=</span> <span class="va">tapseq_genes</span>,  <span class="co"># Compare same genes</span></span>
<span>  TPM_threshold <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; TPM filtering: Kept 19942 out of 19942 genes (threshold: 0.1 TPM)</span></span>
<span><span class="co">#&gt; Gene-specific mode with importance sampling: Using 1000 unique genes, sampled 200 times with weights</span></span>
<span></span>
<span><span class="co"># Extract fold change and expression info for TAP-seq</span></span>
<span><span class="va">fc_expr_tapseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_fc_expression_info.html">extract_fc_expression_info</a></span><span class="op">(</span></span>
<span>  minimum_fold_change <span class="op">=</span> <span class="va">minimum_fold_change</span>,</span>
<span>  gRNA_variability <span class="op">=</span> <span class="va">gRNA_variability</span>,</span>
<span>  custom_pilot_data <span class="op">=</span> <span class="va">tapseq_pilot_data</span>,</span>
<span>  gene_list <span class="op">=</span> <span class="va">tapseq_genes</span>,</span>
<span>  TPM_threshold <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; TPM filtering: Kept 1000 out of 1000 genes (threshold: 0.1 TPM)</span></span>
<span><span class="co">#&gt; Gene-specific mode with importance sampling: Using 1000 unique genes, sampled 200 times with weights</span></span>
<span></span>
<span><span class="co"># Calculate power for perturb-seq</span></span>
<span><span class="va">power_grid_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan_per_grid.html">compute_power_plan_per_grid</a></span><span class="op">(</span></span>
<span>  cells_per_target <span class="op">=</span> <span class="va">cells_per_target</span>,</span>
<span>  reads_per_cell <span class="op">=</span> <span class="va">reads_per_cell</span>,</span>
<span>  fc_expression_df <span class="op">=</span> <span class="va">fc_expr_perturbseq</span><span class="op">$</span><span class="va">fc_expression_df</span>,</span>
<span>  library_parameters <span class="op">=</span> <span class="va">library_params</span>,</span>
<span>  control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>  side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  multiple_testing_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  num_targets <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate power for TAP-seq</span></span>
<span><span class="va">power_grid_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan_per_grid.html">compute_power_plan_per_grid</a></span><span class="op">(</span></span>
<span>  cells_per_target <span class="op">=</span> <span class="va">cells_per_target</span>,</span>
<span>  reads_per_cell <span class="op">=</span> <span class="va">reads_per_cell</span>,</span>
<span>  fc_expression_df <span class="op">=</span> <span class="va">fc_expr_tapseq</span><span class="op">$</span><span class="va">fc_expression_df</span>,</span>
<span>  library_parameters <span class="op">=</span> <span class="va">tapseq_pilot_data</span><span class="op">$</span><span class="va">library_parameters</span>,</span>
<span>  control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>  side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  multiple_testing_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  num_targets <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare results</span></span>
<span><span class="va">comparison_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Technology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Perturb-seq"</span>, <span class="st">"TAP-seq"</span><span class="op">)</span>,</span>
<span>  Genes_Analyzed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fc_expr_perturbseq</span><span class="op">$</span><span class="va">fc_expression_df</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fc_expr_tapseq</span><span class="op">$</span><span class="va">fc_expression_df</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  Overall_Power <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="va">power_grid_ps</span><span class="op">$</span><span class="va">overall_power</span>,</span>
<span>    <span class="va">power_grid_ts</span><span class="op">$</span><span class="va">overall_power</span></span>
<span>  <span class="op">)</span>,</span>
<span>  Cells_per_Target <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="va">power_grid_ps</span><span class="op">$</span><span class="va">cells_per_target</span>,</span>
<span>    <span class="va">power_grid_ts</span><span class="op">$</span><span class="va">cells_per_target</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">comparison_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Technology Genes_Analyzed Overall_Power Cells_per_Target</span></span>
<span><span class="co">#&gt; 1 Perturb-seq            200     0.3373771              100</span></span>
<span><span class="co">#&gt; 2     TAP-seq            200     0.4861346              100</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="power-comparison-across-read-depths">Power Comparison Across Read Depths<a class="anchor" aria-label="anchor" href="#power-comparison-across-read-depths"></a>
</h3>
<p>Now let’s compare how power scales with sequencing depth for both
technologies:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define read depth range</span></span>
<span><span class="va">read_depths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">10000</span>, <span class="fl">15000</span>, <span class="fl">20000</span>, <span class="fl">30000</span>, <span class="fl">40000</span><span class="op">)</span></span>
<span><span class="va">comparison_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">read_depths</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">reads</span> <span class="op">&lt;-</span> <span class="va">read_depths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span></span>
<span>  <span class="co"># Perturb-seq power</span></span>
<span>  <span class="va">power_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan_per_grid.html">compute_power_plan_per_grid</a></span><span class="op">(</span></span>
<span>    cells_per_target <span class="op">=</span> <span class="va">cells_per_target</span>,</span>
<span>    reads_per_cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>    fc_expression_df <span class="op">=</span> <span class="va">fc_expr_perturbseq</span><span class="op">$</span><span class="va">fc_expression_df</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">library_params</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="fl">100</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="co"># TAP-seq power</span></span>
<span>  <span class="va">power_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan_per_grid.html">compute_power_plan_per_grid</a></span><span class="op">(</span></span>
<span>    cells_per_target <span class="op">=</span> <span class="va">cells_per_target</span>,</span>
<span>    reads_per_cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>    fc_expression_df <span class="op">=</span> <span class="va">fc_expr_tapseq</span><span class="op">$</span><span class="va">fc_expression_df</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">tapseq_pilot_data</span><span class="op">$</span><span class="va">library_parameters</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="fl">100</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">comparison_results</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>    Technology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Perturb-seq"</span>, <span class="st">"TAP-seq"</span><span class="op">)</span>,</span>
<span>    Overall_Power <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">power_ps</span><span class="op">$</span><span class="va">overall_power</span>, <span class="va">power_ts</span><span class="op">$</span><span class="va">overall_power</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">comparison_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">comparison_results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot overall power</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">comparison_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">Overall_Power</span>,</span>
<span>                          color <span class="op">=</span> <span class="va">Technology</span>, linetype <span class="op">=</span> <span class="va">Technology</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Technology Comparison: Power vs Sequencing Depth"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%d cells per target, FC = %.2f, FDR = 10%%"</span>,</span>
<span>                      <span class="va">cells_per_target</span>, <span class="va">minimum_fold_change</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Overall Power"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Technology"</span>,</span>
<span>    linetype <span class="op">=</span> <span class="st">"Technology"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>    panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-compare-read-depths-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="cost-benefit-analysis">Cost-Benefit Analysis<a class="anchor" aria-label="anchor" href="#cost-benefit-analysis"></a>
</h3>
<p>Let’s calculate the cost-benefit ratio assuming TAP-seq requires
panel design costs:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assume costs (adjust to your facility's pricing)</span></span>
<span><span class="va">cost_per_million_reads</span> <span class="op">&lt;-</span> <span class="fl">50</span>  <span class="co"># dollars</span></span>
<span><span class="va">tapseq_panel_design_cost</span> <span class="op">&lt;-</span> <span class="fl">5000</span>  <span class="co"># one-time cost</span></span>
<span></span>
<span><span class="co"># Experimental parameters</span></span>
<span><span class="va">num_targets</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">gRNAs_per_target</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="va">total_gRNAs</span> <span class="op">&lt;-</span> <span class="va">num_targets</span> <span class="op">*</span> <span class="va">gRNAs_per_target</span></span>
<span></span>
<span><span class="co"># Calculate total costs for different designs</span></span>
<span><span class="va">cost_comparison</span> <span class="op">&lt;-</span> <span class="va">comparison_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    <span class="co"># Total cells needed for experiment</span></span>
<span>    Total_Cells <span class="op">=</span> <span class="va">Reads_per_Cell</span> <span class="op">*</span> <span class="va">cells_per_target</span> <span class="op">*</span> <span class="va">total_gRNAs</span> <span class="op">/</span> <span class="fl">10</span>,  <span class="co"># MOI = 10</span></span>
<span>    <span class="co"># Total sequencing reads</span></span>
<span>    Total_Reads <span class="op">=</span> <span class="va">Total_Cells</span> <span class="op">*</span> <span class="va">Reads_per_Cell</span>,</span>
<span>    <span class="co"># Sequencing cost</span></span>
<span>    Sequencing_Cost <span class="op">=</span> <span class="op">(</span><span class="va">Total_Reads</span> <span class="op">/</span> <span class="fl">1e6</span><span class="op">)</span> <span class="op">*</span> <span class="va">cost_per_million_reads</span>,</span>
<span>    <span class="co"># Total cost (add panel design for TAP-seq)</span></span>
<span>    Total_Cost <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Technology</span> <span class="op">==</span> <span class="st">"TAP-seq"</span>,</span>
<span>                        <span class="va">Sequencing_Cost</span> <span class="op">+</span> <span class="va">tapseq_panel_design_cost</span>,</span>
<span>                        <span class="va">Sequencing_Cost</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># View cost comparison at 10K reads/cell</span></span>
<span><span class="va">cost_comparison</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Reads_per_Cell</span> <span class="op">==</span> <span class="fl">10000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Technology</span>, <span class="va">Sequencing_Cost</span>, <span class="va">Total_Cost</span>, <span class="va">Overall_Power</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Technology Sequencing_Cost Total_Cost Overall_Power</span></span>
<span><span class="co">#&gt; 1 Perturb-seq           2e+07   20000000     0.3373771</span></span>
<span><span class="co">#&gt; 2     TAP-seq           2e+07   20005000     0.4861346</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h3>
<p>The comparison shows:</p>
<ul>
<li>
<strong>TAP-seq advantages</strong>: Higher power per gene due to
increased sequencing depth on target genes; more efficient use of
sequencing budget for focused studies</li>
<li>
<strong>Perturb-seq advantages</strong>: No panel design cost;
flexibility to analyze any gene post-hoc; better for exploratory
studies</li>
<li>
<strong>Trade-offs</strong>: TAP-seq is cost-effective when you have
well-defined target genes and want maximum power; perturb-seq is better
when target genes are uncertain or for discovery-based experiments</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="comparing-pre-existing-experimental-designs">2. Comparing Pre-existing Experimental Designs<a class="anchor" aria-label="anchor" href="#comparing-pre-existing-experimental-designs"></a>
</h2>
<p>Sometimes you want to compare power between experiments that have
already been performed or are constrained by available resources. Let’s
compare three realistic experimental designs:</p>
<div class="section level3">
<h3 id="define-competing-designs">Define Competing Designs<a class="anchor" aria-label="anchor" href="#define-competing-designs"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Three competing designs with similar total sequencing budget</span></span>
<span><span class="va">designs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Deep"</span>, <span class="st">"Balanced"</span>, <span class="st">"Wide"</span>, <span class="st">"Complete"</span><span class="op">)</span>,</span>
<span>  Cells_per_gRNA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">5000</span>, <span class="fl">10000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  Reads_per_Cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">80000</span>, <span class="fl">40000</span>, <span class="fl">20000</span>, <span class="fl">80000</span><span class="op">)</span>,</span>
<span>  Description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Fewer cells, deeper sequencing"</span>,</span>
<span>    <span class="st">"Balanced cells and depth"</span>,</span>
<span>    <span class="st">"More cells, shallow sequencing"</span>,</span>
<span>    <span class="st">"More cells, deeper sequencing"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">designs</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Design Cells_per_gRNA Reads_per_Cell                    Description</span></span>
<span><span class="co">#&gt; 1     Deep           2000          80000 Fewer cells, deeper sequencing</span></span>
<span><span class="co">#&gt; 2 Balanced           5000          40000       Balanced cells and depth</span></span>
<span><span class="co">#&gt; 3     Wide          10000          20000 More cells, shallow sequencing</span></span>
<span><span class="co">#&gt; 4 Complete          10000          80000  More cells, deeper sequencing</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-power-for-each-design">Calculate Power for Each Design<a class="anchor" aria-label="anchor" href="#calculate-power-for-each-design"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_power_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">designs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span>  <span class="va">power_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_power_plan.html">compute_power_plan</a></span><span class="op">(</span></span>
<span>    baseline_expression_stats <span class="op">=</span> <span class="va">baseline_expression</span>,</span>
<span>    library_parameters <span class="op">=</span> <span class="va">library_params</span>,</span>
<span>    cells_per_target <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Cells_per_gRNA</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    sequenced_reads_per_cell <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Reads_per_Cell</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    MOI <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    num_targets <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    control_group <span class="op">=</span> <span class="st">"complement"</span>,</span>
<span>    side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    prop_non_null <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    multiple_testing_alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    TPM_threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    minimum_fold_change <span class="op">=</span> <span class="fl">0.8</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">design_power_results</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Design <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Design</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    Cells_per_gRNA <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Cells_per_gRNA</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    Num_captured_Cells <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">num_captured_cells</span>,</span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="va">designs</span><span class="op">$</span><span class="va">Reads_per_Cell</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    Overall_Power <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">overall_power</span>,</span>
<span>    Overall_Cost <span class="op">=</span> <span class="fu">perturbplan</span><span class="fu">:::</span><span class="fu"><a href="../reference/cost_computation.html">cost_computation</a></span><span class="op">(</span></span>
<span>      num_captured_cells <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">num_captured_cells</span>,</span>
<span>      sequenced_reads_per_cell <span class="op">=</span> <span class="va">power_result</span><span class="op">$</span><span class="va">sequenced_reads_per_cell</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">design_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">design_power_results</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">design_comparison</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Design Cells_per_gRNA Num_captured_Cells Reads_per_Cell Overall_Power</span></span>
<span><span class="co">#&gt; 1     Deep           2000              20500          80000     0.5522817</span></span>
<span><span class="co">#&gt; 2 Balanced           5000              51250          40000     0.6192765</span></span>
<span><span class="co">#&gt; 3     Wide          10000             102500          20000     0.6430014</span></span>
<span><span class="co">#&gt; 4 Complete          10000             102500          80000     0.7466946</span></span>
<span><span class="co">#&gt;   Overall_Cost</span></span>
<span><span class="co">#&gt; 1      2376.36</span></span>
<span><span class="co">#&gt; 2      5174.20</span></span>
<span><span class="co">#&gt; 3      9581.70</span></span>
<span><span class="co">#&gt; 4     11881.80</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-power-comparison">Visualize Power Comparison<a class="anchor" aria-label="anchor" href="#visualize-power-comparison"></a>
</h3>
<p>Let’s visualize the four designs in the two dimensional space of
overall power and cost:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">design_comparison</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Overall_Cost</span>, y <span class="op">=</span> <span class="va">Overall_Power</span>, color <span class="op">=</span> <span class="va">Design</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Power Comparison of Pre-existing Designs"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Total Cost (USD)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Overall Power"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Design"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>    panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-power-comparison-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="interpretation-1">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-1"></a>
</h3>
<p>The comparison reveals important trade-offs between cell count and
sequencing depth:</p>
<p><strong>Design Characteristics:</strong></p>
<ul>
<li>
<strong>Deep design</strong>: Fewer cells with deeper sequencing
provides higher UMI counts per cell, beneficial for detecting changes in
low-expression genes but limited by smaller sample sizes</li>
<li>
<strong>Balanced design</strong>: Moderate cells and sequencing
depth offers a compromise, balancing detection sensitivity with
statistical power from cell numbers</li>
<li>
<strong>Wide design</strong>: More cells with shallower sequencing
increases statistical power through larger sample sizes but may miss
effects on low-expression genes</li>
<li>
<strong>Complete design</strong>: Maximum cells and depth represents
the gold standard but at significantly higher cost</li>
</ul>
<p><strong>Key Insights:</strong></p>
<ol style="list-style-type: decimal">
<li><p><strong>Power vs Cost</strong>: The “Complete” design achieves
the highest power but at a substantially higher cost than other designs.
The cost increase may not be justified if the “Balanced” or “Wide”
design already achieves acceptable power (e.g., &gt;80%).</p></li>
<li><p><strong>Diminishing Returns</strong>: Moving from “Wide” to
“Complete” shows diminishing returns in power relative to cost increase,
suggesting that moderate designs often provide the best value.</p></li>
<li>
<p><strong>Expression-Dependent Effects</strong>:</p>
<ul>
<li>Low-expression genes benefit more from the “Deep” design due to
higher UMI counts reducing sampling noise</li>
<li>High-expression genes benefit more from the “Wide” design since
additional cells provide more statistical power than marginal UMI
increases</li>
</ul>
</li>
<li><p><strong>Experimental Constraints</strong>: Real experiments often
face constraints (e.g., limited cells from rare cell types, fixed
sequencing budget). This comparison helps identify which constraint
matters most for your specific application.</p></li>
</ol>
<p><strong>Practical Recommendations:</strong></p>
<ul>
<li>If most target genes are lowly expressed (&lt; 10 TPM): favor deeper
sequencing designs</li>
<li>If target genes span a wide expression range: favor balanced or wide
designs to ensure adequate power across all genes</li>
<li>If discovery is the goal: favor wide designs to maximize the number
of cells and improve rare cell type detection</li>
<li>If validating specific hypotheses: consider deep designs to maximize
sensitivity for target genes</li>
</ul>
<p>The optimal choice depends on your biological question, expression
range of target genes, and budget constraints. Use the saturation curves
(Section 3) to identify the point of diminishing returns for sequencing
depth in your pilot data.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-expression-distributions-and-saturation-curves">3. Visualizing Expression Distributions and Saturation Curves<a class="anchor" aria-label="anchor" href="#visualizing-expression-distributions-and-saturation-curves"></a>
</h2>
<p>Understanding the underlying statistical models helps interpret power
analysis results.</p>
<div class="section level3">
<h3 id="gene-expression-distribution">Gene Expression Distribution<a class="anchor" aria-label="anchor" href="#gene-expression-distribution"></a>
</h3>
<p>Let’s visualize the distribution of gene expression levels in the
pilot data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare expression data</span></span>
<span><span class="va">expression_df</span> <span class="op">&lt;-</span> <span class="va">baseline_expression</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    TPM <span class="op">=</span> <span class="va">relative_expression</span> <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span>    Log10_TPM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">TPM</span> <span class="op">+</span> <span class="fl">0.1</span><span class="op">)</span>  <span class="co"># Add pseudocount for log</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Histogram of TPM values</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">expression_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TPM</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span>, fill <span class="op">=</span> <span class="st">"steelblue"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Distribution of Gene Expression Levels"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%d genes in baseline data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">expression_df</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Expression Level (TPM)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Number of Genes"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-expression-distribution-1.png" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Expression vs dispersion relationship</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">expression_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TPM</span>, y <span class="op">=</span> <span class="va">expression_size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, color <span class="op">=</span> <span class="st">"red"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Gene Expression vs Dispersion Parameter"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="st">"Lower dispersion = higher biological variability"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Expression Level (TPM)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Dispersion Size Parameter"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-expression-distribution-2.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="saturation-curve-visualization">Saturation Curve Visualization<a class="anchor" aria-label="anchor" href="#saturation-curve-visualization"></a>
</h3>
<p>The saturation curve relates reads per cell to UMIs per cell. Let’s
visualize this relationship:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract library parameters</span></span>
<span><span class="va">UMI_max</span> <span class="op">&lt;-</span> <span class="va">library_params</span><span class="op">$</span><span class="va">UMI_per_cell</span></span>
<span><span class="va">variation</span> <span class="op">&lt;-</span> <span class="va">library_params</span><span class="op">$</span><span class="va">variation</span></span>
<span></span>
<span><span class="co"># Generate saturation curve</span></span>
<span><span class="va">reads_range</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">100000</span>, length.out <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">umis_predicted</span> <span class="op">&lt;-</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">reads_range</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                               <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="va">reads_range</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">saturation_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Reads_per_Cell <span class="op">=</span> <span class="va">reads_range</span>,</span>
<span>  UMIs_per_Cell <span class="op">=</span> <span class="va">umis_predicted</span>,</span>
<span>  Saturation_Pct <span class="op">=</span> <span class="op">(</span><span class="va">umis_predicted</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot saturation curve</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">saturation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">UMIs_per_Cell</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">UMI_max</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">80000</span>, y <span class="op">=</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="fl">1.05</span>,</span>
<span>           label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Max UMI = %d"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">UMI_max</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"red"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Add example design points</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span>,</span>
<span>    Design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Wide"</span>, <span class="st">"Balanced"</span>, <span class="st">"Deep"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">Reads_per_Cell</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                                              <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="va">Reads_per_Cell</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Reads_per_Cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span>,</span>
<span>    Design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Wide"</span>, <span class="st">"Balanced"</span>, <span class="st">"Deep"</span><span class="op">)</span>,</span>
<span>    UMI <span class="op">=</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                       <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">UMI</span>, label <span class="op">=</span> <span class="va">Design</span><span class="op">)</span>,</span>
<span>  vjust <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, color <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Read-to-UMI Saturation Curve"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"UMI_per_cell = %d, variation = %.3f"</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">UMI_max</span><span class="op">)</span>, <span class="va">variation</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"UMIs per Cell"</span>,</span>
<span>    caption <span class="op">=</span> <span class="st">"Green points show example experimental designs"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-saturation-curve-1.png" width="100%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Saturation percentage vs reads</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">saturation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">Saturation_Pct</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">80</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"orange"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">90</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">90000</span>, y <span class="op">=</span> <span class="fl">82</span>, label <span class="op">=</span> <span class="st">"80% saturation"</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"orange"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">90000</span>, y <span class="op">=</span> <span class="fl">92</span>, label <span class="op">=</span> <span class="st">"90% saturation"</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"red"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Sequencing Saturation vs Read Depth"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Saturation (%)"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-saturation-curve-2.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="marginal-utility-of-additional-reads">Marginal Utility of Additional Reads<a class="anchor" aria-label="anchor" href="#marginal-utility-of-additional-reads"></a>
</h3>
<p>As sequencing depth increases, the marginal benefit decreases due to
saturation. Let’s visualize this:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate marginal UMIs per additional 1000 reads</span></span>
<span><span class="va">saturation_df</span> <span class="op">&lt;-</span> <span class="va">saturation_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    Marginal_UMI <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">UMIs_per_Cell</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">Reads_per_Cell</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">saturation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Reads_per_Cell</span>, y <span class="op">=</span> <span class="va">Marginal_UMI</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0.5</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html" class="external-link">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">80000</span>, y <span class="op">=</span> <span class="fl">0.55</span>,</span>
<span>           label <span class="op">=</span> <span class="st">"&lt; 0.5 UMI per 1000 reads: diminishing returns"</span>,</span>
<span>           color <span class="op">=</span> <span class="st">"red"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Marginal Utility of Additional Sequencing"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="st">"Additional UMIs gained per 1000 extra reads"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Reads per Cell"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Marginal UMIs per 1000 Reads"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-marginal-utility-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="simulating-expression-for-individual-cells">Simulating Expression for Individual Cells<a class="anchor" aria-label="anchor" href="#simulating-expression-for-individual-cells"></a>
</h3>
<p>Let’s simulate what gene expression looks like for individual cells
at different sequencing depths:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select a few example genes across expression range</span></span>
<span><span class="va">example_genes</span> <span class="op">&lt;-</span> <span class="va">baseline_expression</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">relative_expression</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">500</span>, <span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Gene_Label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene "</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                            <span class="st">" ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">relative_expression</span> <span class="op">*</span> <span class="fl">1e6</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">" TPM)"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate cells at different read depths</span></span>
<span><span class="va">simulate_expression</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gene_data</span>, <span class="va">n_cells</span>, <span class="va">reads_per_cell</span>, <span class="va">library_params</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Calculate UMIs per cell from reads</span></span>
<span>  <span class="va">UMI_max</span> <span class="op">&lt;-</span> <span class="va">library_params</span><span class="op">$</span><span class="va">UMI_per_cell</span></span>
<span>  <span class="va">variation</span> <span class="op">&lt;-</span> <span class="va">library_params</span><span class="op">$</span><span class="va">variation</span></span>
<span>  <span class="va">umi_per_cell</span> <span class="op">&lt;-</span> <span class="va">UMI_max</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">reads_per_cell</span> <span class="op">/</span> <span class="va">UMI_max</span><span class="op">)</span> <span class="op">*</span></span>
<span>                               <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">variation</span> <span class="op">*</span> <span class="va">reads_per_cell</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">UMI_max</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Simulate gene expression using negative binomial</span></span>
<span>  <span class="co"># mean = library_size * relative_expression</span></span>
<span>  <span class="co"># size = expression_size</span></span>
<span>  <span class="va">gene_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span></span>
<span>    n <span class="op">=</span> <span class="va">n_cells</span>,</span>
<span>    mu <span class="op">=</span> <span class="va">umi_per_cell</span> <span class="op">*</span> <span class="va">gene_data</span><span class="op">$</span><span class="va">relative_expression</span>,</span>
<span>    size <span class="op">=</span> <span class="va">gene_data</span><span class="op">$</span><span class="va">expression_size</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">gene_expression</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Simulate for different read depths (matching Section 2 designs)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">simulation_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">example_genes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">reads</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20000</span>, <span class="fl">40000</span>, <span class="fl">80000</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">gene_counts</span> <span class="op">&lt;-</span> <span class="fu">simulate_expression</span><span class="op">(</span></span>
<span>      <span class="va">example_genes</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span>,</span>
<span>      n_cells <span class="op">=</span> <span class="fl">500</span>,</span>
<span>      reads_per_cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>      library_params <span class="op">=</span> <span class="va">library_params</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="va">simulation_results</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">simulation_results</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>      Gene <span class="op">=</span> <span class="va">example_genes</span><span class="op">$</span><span class="va">Gene_Label</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>      Reads_per_Cell <span class="op">=</span> <span class="va">reads</span>,</span>
<span>      Expression <span class="op">=</span> <span class="va">gene_counts</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">simulation_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">simulation_results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot distributions</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">simulation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Expression</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">Reads_per_Cell</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.6</span>, bins <span class="op">=</span> <span class="fl">30</span>, position <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">Gene</span>, scales <span class="op">=</span> <span class="st">"free"</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span>,</span>
<span>                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"20K (Wide)"</span>, <span class="st">"40K (Balanced)"</span>, <span class="st">"80K (Deep)"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Simulated Cell-to-Cell Expression Distributions"</span>,</span>
<span>    subtitle <span class="op">=</span> <span class="st">"500 cells per condition across three experimental designs"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"UMI Counts"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Number of Cells"</span>,</span>
<span>    fill <span class="op">=</span> <span class="st">"Design"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignettes%2Ffigures%2Fprospective-power-simulate-cells-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="interpretation-2">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation-2"></a>
</h3>
<p>The visualizations reveal key insights about the underlying
statistical models:</p>
<p><strong>Expression Distribution:</strong> - Most genes have
low-to-medium expression (&lt; 100 TPM), with a heavy right tail of
highly expressed genes - The dispersion parameter (expression_size)
tends to be lower for lowly expressed genes, indicating higher
biological variability - This heterogeneity means different genes
benefit from different experimental designs</p>
<p><strong>Saturation Effects:</strong> - The saturation curve shows
diminishing returns beyond ~40,000 reads per cell for this dataset - At
20,000 reads (Wide design): moderate saturation, cost-effective for most
applications - At 40,000 reads (Balanced design): approaching
diminishing returns, suitable for comprehensive studies - At 80,000
reads (Deep design): significant saturation, only justified for
detecting subtle changes in low-expression genes</p>
<p><strong>Marginal Utility:</strong> - Each additional 1,000 reads
provides progressively fewer additional UMIs - Beyond 40,000-50,000
reads per cell, marginal utility drops below 0.5 UMIs per 1,000 reads -
This explains why the “Complete” design in Section 2 shows diminishing
returns relative to its cost</p>
<p><strong>Cell-Level Variation:</strong> - The simulated distributions
show that: - Low-expression genes (&lt; 10 TPM) have high coefficient of
variation, making them harder to detect without deeper sequencing -
High-expression genes show more consistent expression across cells,
benefiting more from additional cells than additional depth - The three
design points (20K, 40K, 80K reads/cell) show how sequencing depth
affects detection resolution</p>
<p><strong>Connection to Section 2 Designs:</strong> - <strong>Wide
design (20K reads)</strong>: Sits in the steep part of the saturation
curve, maximizing cells while maintaining reasonable UMI counts -
<strong>Balanced design (40K reads)</strong>: Reaches the inflection
point where marginal returns start to decline noticeably - <strong>Deep
design (80K reads)</strong>: Well into the saturation plateau, justified
only for specific applications requiring maximum sensitivity for
low-expression targets</p>
</div>
</div>
<div class="section level2">
<h2 id="practical-recommendations">Practical Recommendations<a class="anchor" aria-label="anchor" href="#practical-recommendations"></a>
</h2>
<p>Based on these analyses:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Technology choice</strong>:
<ul>
<li>Use TAP-seq when you have a well-defined gene panel and want maximum
power per gene</li>
<li>Use perturb-seq for exploratory studies or when target genes are
uncertain</li>
</ul>
</li>
<li>
<strong>Design optimization</strong>:
<ul>
<li>Balance cells vs depth based on your expression range of
interest</li>
<li>Consider saturation curves when deciding sequencing depth</li>
<li>Low-expression genes benefit more from deeper sequencing</li>
<li>High-expression genes benefit more from additional cells</li>
</ul>
</li>
<li>
<strong>Cost considerations</strong>:
<ul>
<li>Calculate cost per expected discovery, not just total cost</li>
<li>Account for fixed costs (panel design) vs variable costs
(sequencing)</li>
<li>Consider pilot experiments to refine parameters before large
studies</li>
</ul>
</li>
<li>
<strong>Model validation</strong>:
<ul>
<li>Visualize your pilot data’s expression distribution and saturation
curve</li>
<li>Check that model assumptions (NB distribution, saturation curve) fit
your data</li>
<li>If fit is poor, consider collecting additional pilot data</li>
</ul>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 Patched (2024-06-25 r86831)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS 15.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] tidyr_1.3.1       dplyr_1.1.4       ggplot2_3.5.2     perturbplan_0.1.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.7-0       gtable_0.3.6       jsonlite_2.0.0     compiler_4.4.1    </span></span>
<span><span class="co">#&gt;  [5] tidyselect_1.2.1   Rcpp_1.1.0         minpack.lm_1.2-4   jquerylib_0.1.4   </span></span>
<span><span class="co">#&gt;  [9] splines_4.4.1      scales_1.4.0       systemfonts_1.1.0  textshaping_0.4.0 </span></span>
<span><span class="co">#&gt; [13] yaml_2.3.10        fastmap_1.2.0      lattice_0.22-6     R6_2.6.1          </span></span>
<span><span class="co">#&gt; [17] labeling_0.4.3     generics_0.1.4     knitr_1.50         htmlwidgets_1.6.4 </span></span>
<span><span class="co">#&gt; [21] tibble_3.3.0       desc_1.4.3         RColorBrewer_1.1-3 bslib_0.9.0       </span></span>
<span><span class="co">#&gt; [25] pillar_1.11.0      R.utils_2.12.3     rlang_1.1.6        cachem_1.1.0      </span></span>
<span><span class="co">#&gt; [29] xfun_0.52          fs_1.6.6           sass_0.4.10        cli_3.6.5         </span></span>
<span><span class="co">#&gt; [33] mgcv_1.9-1         withr_3.0.2        pkgdown_2.1.1      magrittr_2.0.3    </span></span>
<span><span class="co">#&gt; [37] digest_0.6.37      grid_4.4.1         rstudioapi_0.16.0  nlme_3.1-166      </span></span>
<span><span class="co">#&gt; [41] lifecycle_1.0.4    R.methodsS3_1.8.2  R.oo_1.26.0        vctrs_0.6.5       </span></span>
<span><span class="co">#&gt; [45] evaluate_1.0.4     glue_1.8.0         data.table_1.17.8  farver_2.1.2      </span></span>
<span><span class="co">#&gt; [49] ragg_1.3.3         purrr_1.1.0        rmarkdown_2.29     tools_4.4.1       </span></span>
<span><span class="co">#&gt; [53] pkgconfig_2.0.3    htmltools_0.5.8.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="see-also">See Also<a class="anchor" aria-label="anchor" href="#see-also"></a>
</h2>
<ul>
<li>
<code><a href="../reference/compute_power_plan.html">?compute_power_plan</a></code>: Prospective power analysis for
experimental planning</li>
<li>
<code><a href="../reference/compute_power_plan_per_grid.html">?compute_power_plan_per_grid</a></code>: Grid-based power
analysis</li>
<li>
<code><a href="../reference/library_estimation.html">?library_estimation</a></code>: Library parameter estimation from
QC data</li>
<li>Vignette “Preprocess Reference Expression Data”: Creating custom
pilot data</li>
<li>Vignette “Getting Started with PerturbPlan”: Basic power analysis
workflow</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ziang Niu, Yihui He, Eugene Katsevich.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
